watchface

```diff
@@ -0,0 +1,406 @@
+            if [ "$ADB" = "true" ] ; then
+                printf -v cmd %q "${cmd}"
+                adb shell "su -l -c ${cmd} ceres"
+            else
+                ssh -p "${WATCHPORT}" -t ceres@"${WATCHADDR}" ${cmd}
+            fi
+            ;;
+        *)
+            echo "Error: unknown watch user ${user}"
+            ;;
+    esac
+}
+
+function setDconf {
+    local dconfsetting="$1"
+    local filename="$2"
+    doWatchCommand "ceres" "dconf write ${dconfsetting} '\"file://${filename}\"'"
+}
+
+function pushFiles {
+    local user="$1"
+    local sourcedir="$2"
+    local destdir="$3"
+    if [ "$ADB" = "true" ] ; then
+        adb push ${sourcedir} "${destdir}"
+    else
+        scp -P"${WATCHPORT}" -r ${sourcedir} "${user}@${WATCHADDR}:${destdir}"
+    fi
+}
+
+function pullFiles {
+    local user="$1"
+    local sourcedir="$2"
+    local destdir="$3"
+    if [ "$ADB" = "true" ] ; then
+        adb pull ${sourcedir} "${destdir}"
+    else
+        scp -P"${WATCHPORT}" -r "${user}@${WATCHADDR}:${sourcedir}" "${destdir}"
+    fi
+}
+
+function pushWatchface {
+    pushFiles "root" "${1}"'/usr/share/*' "/usr/share/"
+}
+
+function pushWallpaper {
+    local source="$1"
+    local wallpaper
+    wallpaper="$(basename "$1")"
+    local destination="/usr/share/asteroid-launcher/wallpapers/full/${wallpaper}"
+    pushFiles "root" "${source}" "${destination}"
+    setDconf "/desktop/asteroid/background-filename" "${destination}"
+}
+
+function restartCeres {
+    doWatchCommand "root" "systemctl restart user@1000"
+}
+
+function rebootWatch {
+    doWatchCommand "ceres" "reboot"
+}
+
+function activateWatchface {
+    setDconf "/desktop/asteroid/watchface" "/usr/share/asteroid-launcher/watchfaces/${1}.qml"
+}
+
+function setTimeZone {
+    doWatchCommand "root" "timedatectl set-timezone $1"
+}
+
+function forgetKeys {
+    ssh-keygen -R "${WATCHADDR}"
+    ssh-keygen -R watch
+}
+
+function watchfaceClone {
+    local files
+    mapfile -t files <<<"$(find . -path './'"$1"'/*' -type f | awk '{ dst=$0 ; gsub(/'"$1"'/,"mmmm",dst);  print $0 " " dst }')"
+    for myfile in "${files[@]}" ; do
+        IFS=" " read -r -a onefile <<< "${myfile}"
+        #echo "Copying from ${onefile[0]} to ${onefile[1]}"
+        destdir="${onefile[1]%/*}"
+        extension="${onefile[1]##*.}"
+        mkdir -p "${destdir}"
+        cp "${onefile[0]}" "${onefile[1]}"
+        if [ "${extension}" = "qml" ] ; then
+            sed -i "s/$1/$2/g" "${onefile[1]}"
+        fi
+    done
+}
+
+function deployface {
+    # trim trailing / if present
+    local face="${1%/*}"
+    local activate="$2"
+    if [ -d "${face}/usr/share/asteroid-launcher/watchfaces/" ] ; then
+        echo "Deploying ${face}"
+        pushWatchface "${face}"
+        if [ "${activate}" = "true" ] ; then
+            activateWatchface "${face}"
+            if [ -n "${WALLPAPER}" ] ; then
+                pushWallpaper "${WALLPAPER}"
+            fi
+            restartCeres
```

Comment:
If I understand the documentation correctly, restarting the session is necessary to trigger a font cache recalculation.
If the user installs multiple watchfaces all at once and "activate" is only set to true for the first one, that means that asteroid-launcher will be restarted after the first watchface install and all remaining watchfaces will have their fonts not calculated.
It feels like restarting ceres should be done in a similar place as rebooting the watch.

Reply:
Good point. I've moved the restart to after the last watch is pushed.